etl/steps/data/meadow/climate/2024-11-19/total_precipitation.py

```diff
@@ -0,0 +1,167 @@
+# Initialize logger.
+log = get_logger()
+
+
+def _load_data_array(snap: Snapshot) -> xr.DataArray:
+    log.info("load_data_array.start")
+    # Load data from snapshot.
+    with zipfile.ZipFile(snap.path, "r") as zip_file:
+        # Iterate through all files in the zip archive
+        for file_info in zip_file.infolist():
+            with zip_file.open(file_info) as file:
+                file_content = file.read()
+    # Create an in-memory bytes file and load the dataset
+    with io.BytesIO(file_content) as memfile:
+        da = xr.open_dataset(memfile).load()  # .load() ensures data is eagerly loaded
+    da = da["tp"]
+    # Set the coordinate reference system for the precipitation data to EPSG 4326.
+    da = da.rio.write_crs("epsg:4326")
+
+    return da
+
+
+def _load_shapefile(file_path: str) -> gpd.GeoDataFrame:
+    log.info("load_shapefile.start")
+    shapefile = gpd.read_file(file_path)
+    return shapefile[["geometry", "WB_NAME"]]
+
+
+def run(dest_dir: str) -> None:
+    # Activates the usage of the global context. Using this option can enhance the performance
+    # of initializing objects in single-threaded applications.
+    pyproj.set_use_global_context(True)  # type: ignore
+
+    #
+    # Load inputs.
+    #
+    # Retrieve snapshot.
+    snap = paths.load_snapshot("total_precipitation.zip")
+
+    # Read data from snapshot
+    da = _load_data_array(snap)
+
+    # Read the shapefile to extract country informaiton
+    snap_geo = paths.load_snapshot("world_bank.zip")
+    shapefile_name = "WB_countries_Admin0_10m/WB_countries_Admin0_10m.shp"
+
+    # Check if the shapefile exists in the ZIP archive
+    with zipfile.ZipFile(snap_geo.path, "r"):
+        # Construct the correct path for Geopandas
+        file_path = f"zip://{snap_geo.path}!/{shapefile_name}"
+
+        # Read the shapefile directly from the ZIP archive
+        shapefile = _load_shapefile(file_path)
+
+    #
+    # Process data.
+    #
+
+    # Initialize an empty dictionary to store the country-wise average precipitation.
+    temp_country = {}
+
+    # Add Global mean precipitation
+    weights = np.cos(np.deg2rad(da.latitude))
+    weights.name = "weights"
+    clim_month_weighted = da.weighted(weights)
+    global_mean = clim_month_weighted.mean(["longitude", "latitude"])
+    temp_country["World"] = global_mean
+
+    # Initialize a list to keep track of small countries where precipitation data extraction fails.
+    small_countries = []
+
+    # Iterate over each row in the shapefile data.
+    for i in tqdm(range(shapefile.shape[0])):
+        # Extract the data for the current row.
+        geometry = shapefile.iloc[i]["geometry"]
+        country_name = shapefile.iloc[i]["WB_NAME"]
+
+        try:
+            # Clip to the bounding box for the country's shape to significantly improve performance.
+            xmin, ymin, xmax, ymax = geometry.bounds
+            clip = da.rio.clip_box(minx=xmin, miny=ymin, maxx=xmax, maxy=ymax)
```

Comment:
Hi! Chile doesn't look bad. I would like to know more about cases like Croatia, Brazil, or Italy, where the clip probably considers whole other countries.

Reply:
hi :) thanks a lot for a thoughtful review btw! as I note in the limitations of the data the problem only really appears once you go down to small countries and I think no matter what shapes we pick we will ultimately run into this problem. I am attaching the screenshots of the countries you've wanted to see below though just to highlight this point further - e.g., Croatia is the worst but largely because it's the smallest
<img width="512" alt="Screenshot 2024-11-21 at 18 43 26" src="https://github.com/user-attachments/assets/24df76b8-db25-4114-8a30-b1e10252d5e0">
<img width="491" alt="Screenshot 2024-11-21 at 18 44 37" src="https://github.com/user-attachments/assets/2d3270d5-13cf-4770-aa07-62236c913d41">
<img width="526" alt="Screenshot 2024-11-21 at 18 43 53" src="https://github.com/user-attachments/assets/9d8be6f9-d03a-453f-96fe-e1f910ffd529">

Reply:
So the clip method captures all these small squares inside the country, but you use the box to simplify the limits? At first I thought the clipping for Brazil would make you calculate data for the entire territory of Paraguay and Bolivia, for example.

If that's not the case we are fine

Reply:
yep, sorry I definitely double checked and this does not produce this issue. It just helps the code to avoid running into memory issues because you first find an approximate location of the country and then use the exact boundaries later in the code. So instead of loading the whole dataset and searching for the boundaries within the whole globe you search within a smaller space. The actual clipping to extract the data happens here` clip = clip.rio.clip([mapping(geometry)], shapefile.crs)
`